### PR TITLE
Improve CropMirrorNormalize operator performance

### PR DESCRIPTION
**Pull request by jmicota**
## Category:

**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:

Work's main motivation was to improve throughput for small batch sizes of data.

Change improves throughput of CropMirrorNormalize operator for all batch sizes, measured in GB/s. Below shown plotted results of improvements for different batch sizes.

- _CMN_original_ - current behaviour of CropMirrorNormalize
- _CMN_block_256_ - throughput with change of kBlockDim from 512 to 256 
- _CMN_blockSize_sm_ - throughput with change of _CMN_block_256_ and change of kBlockSize to adaptive values, described below in detail
- _CMN_merged_copy_ - throughput with changes of _CMN_block_256_ and _CMN_blockSize_sm_ and merge of two seperate calls to memcpy into one

image

image

**Detailed change description:**
- Experimentally found new value of kBlockDim through plotting throughput with many different values of kBlockDim using existing CropMirrorNormalize benchmark
- Out of three separate calls to copy from Host to Device in slice_flip_normal_permute_pad_gpu.h two were merged into one resulting in slightly higher throughput for small batch sizes
- **Adaptive value of kBlockSize:**

   Nvidia Nsight Compute measurement results for original CropMirrorNormalize kernel:
--- **0,3 full waves** across all SMs
--- **24,03 %** of achieved occupancy (max 75%)
--- **76 blocks** launched (80 SMs available, limit of 6 blocks per SM)

   Nvidia Nsight Compute measurement results for improved CropMirrorNormalize kernel (measurements include all improvements from this PR):
--- **1 full wave** across all SMs
--- **53,55 %** of achieved occupancy (max 75%)
--- **480 blocks** launched (80 SMs available, limit of 6 blocks per SM, so value of 80*6)

   Value of kBlockSize is being set to such that there will be number of blocks divisible by **x** = (number of SMs times block limit per SM). Because number of blocks is found based on blockSize, first we count the blocks as in original CropMirrorNormalize code with kBlockSize set to 32 (originally 64, 32 behaved better in testing different values of kBlockDim), then we round block_count up to become divisible by **x**, and finally we adjust kBlockSize to count the blocks once more, now based on new kBlockSize.

### Affected modules and functionalities:

slice_flip_normalize_pad_gpu.h

### Tests
- [x] Existing tests apply
- [ ] New tests added
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h [hunk 1]
```diff
@@ -101,13 +102,27 @@ class SliceFlipNormalizePermutePadGpu {
     se.add<mm::memory_kind::pinned, detail::SampleDesc<Dims>>(num_samples);
     se.add<mm::memory_kind::device, detail::SampleDesc<Dims>>(num_samples);
 
+    int blocks_per_sm_;
+    CUDA_CALL(cudaOccupancyMaxActiveBlocksPerMultiprocessor(&blocks_per_sm_,
+      detail::SliceFlipNormalizePermutePadKernel<false, false, false, OutputType, InputType, Dims>,
```
Comment (JanuszL) [inline dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:107]: Is there any difference (significant) if the occupancy is calculated for `detail::SliceFlipNormalizePermutePadKernel<false, false, false` but different variant is used in the bechmark?

Comment (jmicota) [reply dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:107]: There is no difference for this call, it always returns 6 blocks per SM.

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h [hunk 2]
```diff
@@ -165,8 +175,12 @@ class SliceFlipNormalizePermutePadGpu {
       for (int d = 0; d < nfill_values_; d++)
         fill_values[d] = sample_args.fill_values[d];
     }
-    OutputType *fill_values_gpu = context.scratchpad->ToGPU(
-        context.gpu.stream, make_span(fill_values_cpu, num_samples * nfill_values_));
+    OutputType *fill_values_gpu;
+    std::tie(norm_add_gpu, norm_mul_gpu, fill_values_gpu) = context.scratchpad->ToContiguousGPU(
```
Comment (JanuszL) [inline dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:179]: This is probably a long shot but what would be the difference if you allocate `norm_add_cpu`, `norm_mul_cpu`, and `fill_values_cpu` as one continuous block and then copy it as such (it should be just one copy invocation), but it would require manual pointer math.

Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:179]: ToContiguousGPU will result in a single h2d copy. What we would have here are extra h2h copies to a pinned staging buffer. I think that we can live with this (very small amount of data copied). I wouldn't complicate this with manual pointer math.

What we could change is that the buffers for norm_add_cpu and norm_mul_cpu don't need to be pinned (since they are copied to a pinned staging buffer by ToContiguousGPU):

Example, we could do `AllocateHost` instead of `AllocatePinned`:
```
norm_add_cpu = context.scratchpad->AllocateHost<float>(num_samples * norm_args_size_);
```

We can always take care of this in a follow-up PR, as it is a little bit out of scope here.

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
Build failed on some linter problems:
```
+ make -j lint
Scanning dependencies of target lint
Performing linter check
/opt/dali/dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:107:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
/opt/dali/dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:108:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
/opt/dali/dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:125:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
/opt/dali/dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:180:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
```

**Comment by szalpal** [comment]
!build

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h (hunk 2 above)
Comment (JanuszL) [reply dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:179]: Agree

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h (hunk 1 above)
Comment (JanuszL) [reply dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h]: Understood

**Review by JanuszL — APPROVED** [review]
Very nice PR. Good work.

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h [hunk 3]
```diff
@@ -101,12 +102,26 @@ class SliceFlipNormalizePermutePadGpu {
...
       size_t sample_size = volume(elem.shape);
+      all_sample_sizes += sample_size;
       block_count_ += std::ceil(
         sample_size / static_cast<float>(kBlockSize));
     }
+    if (block_count_ % number_of_blocks != 0) {
+      block_count_ += (number_of_blocks - (block_count_ % number_of_blocks));
+    }
+    kBlockSize = div_ceil(all_sample_sizes, block_count_);
+    block_count_ = std::ceil(all_sample_sizes / static_cast<float>(kBlockSize));
```
Comment (szalpal) [inline dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:124]: Clang-tidy would probably suggest adding `static_cast` here, since `std::ceil` returns float and `block_count_` is `size_t`. Please consider.

**Review by szalpal — APPROVED** [review]
Cool PR! Please consider a few nitpicks.

**Comment by szalpal** [comment]
CI tests failed on:
```
test_operator_crop_mirror_normalize.py
test_dali_variable_batch_size.py
```
Please try to run those locally and see if you can tackle them this way. I'll send the logs via another channel.

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h [hunk 4]
```diff
@@ -101,12 +102,26 @@ class SliceFlipNormalizePermutePadGpu {
...
+    size_t all_sample_sizes = 0;
     for (auto &elem : args) {
       size_t sample_size = volume(elem.shape);
+      all_sample_sizes += sample_size;
       block_count_ += std::ceil(
         sample_size / static_cast<float>(kBlockSize));
     }
+    if (block_count_ % number_of_blocks != 0) {
+      block_count_ += (number_of_blocks - (block_count_ % number_of_blocks));
+    }
```
[the remark this answers is not in the document]
Comment (szkarpinski) [reply dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:122]: Nitpicking in the nitpick: probably this should be `remainder`, not `reminder`

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h (hunk 3 above)
Comment (szkarpinski) [inline dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:124]: I think the bug that causes the tests to fail is in this line.

It'll be probably easiest to explain it by example. Each block can work on only one sample.  Let's say we have 10 samples, each of size 100, and kBlockSize is 1000. We need 10 blocks, because each block can work on only one sample. But `all_sample_sizes / kBlockSize` is `1`.

Calculating `block_count_` as a sum over the samples seems to do the trick:

```suggestion
    block_count_ = 0;
    for (auto &elem : args) {
      block_count_ += std::ceil(volume(elem.shape) / static_cast<float>(kBlockSize));
    }```

Comment (szkarpinski) [inline dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:124]: Also the tests seem not to catch it, but we'll have division by zero here if all samples have size 0. Not sure if it's a real problem though.

**Comment by szalpal** [comment]
!build
